### ISSUE TITLE
Potential Bug in Configuration when MBEDTLS_ECDSA_SIGN_ALT is enabled

### ISSUE THREAD

**Issue by umanayana** [ARMmbed/mbedtls#2009]
Hi, 
I am adding a custom ECDSA signature process thus I enable the configuration flag `MBEDTLS_ECDSA_SIGN_ALT`.

When I do this the method `mbedtls_ecdsa_sign()` is removed. I then write code which setups my own `mbedtls_pk_context` with custom callback to signing method and so on.

What I've noticed is the method `mbedtls_ecdsa_write_signature()` calls `mbedtls_ecdsa_sign()`. This `mbedtls_ecdsa_write_signature()` would never be called and should be disabled in some manner.
It never gets called because  `mbedtls_pk_context::pk_info` actually points to a custom `pk_info` which uses a custom `sign_func` method.

I've added the following to `mbedtls_ecdsa_write_signature()` method.

```
#if defined(MBEDTLS_ECDSA_DETERMINISTIC)
    (void) f_rng;
    (void) p_rng;

    MBEDTLS_MPI_CHK( mbedtls_ecdsa_sign_det( &ctx->grp, &r, &s, &ctx->d,
                             hash, hlen, md_alg ) );
#elif !defined(MBEDTLS_ECDSA_SIGN_ALT)
    (void) md_alg;

    MBEDTLS_MPI_CHK( mbedtls_ecdsa_sign( &ctx->grp, &r, &s, &ctx->d,
                         hash, hlen, f_rng, p_rng ) );
#else
    ret = MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
    goto cleanup;
#endif
```

**Comment by Patater** [comment]
Hi @umanayana 

Thanks for the information about your experience using the `MBEDTLS_ECDSA_SIGN_ALT` feature.

Are you proposing a change to `mbedtls_ecdsa_write_signature()`? If so, could you please open a PR?

Do you have some specific question about your experience?

Thanks!

**Comment by RonEld** [comment]
Hi @umanayana 
The purpose of `mbedtls_ecdsa_write_signature()` is to sign the data, and then encode it in ASN1 notation, for usage of TLS, for example.
You are suggesting to have this function return `MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE` , in case there is an alternative implementation for ecdsa signing. This will cause TLS to fail in such platforms that have alternative ecdsa signing.

In your specific use case, you probably don't need this function, so I suggest having the linker remove function from the binary image.

Does this makes sense?

In addition, note that `MBEDTLS_ECDSA_SIGN_ALT ` does not use a different `mbedtls_pk_context` as it doesn't receives this context as parameter.
Your change is vert spec ific to your use case, and I don't see how it can be upstreamed.

May I know why you modified your own `pk_info` ?

**Comment by umanayana** [comment]
My post is probably not clear. 

The API does not properly support replacing the ECDSA signature process properly. Especially when you want to use a hardware accelerator.

I had adjusted the API to allow for better support. Here is what I have done.

The `mbedtls_ecdsa_info` variable ( which is a `mbedtls_pk_info_t` structure) provides all the mappings for the callback. The callback in here is what should be stubbed out.

**pk_wrap.c**

Wrap the 
```
#if !defined(MBEDTLS_ECDSA_SIGN_ALT)
static int ecdsa_sign_wrap( void *ctx, mbedtls_md_type_t md_alg,
                   const unsigned char *hash, size_t hash_len,
                   unsigned char *sig, size_t *sig_len,
                   int (*f_rng)(void *, unsigned char *, size_t), void *p_rng )
{
    return( mbedtls_ecdsa_write_signature( (mbedtls_ecdsa_context *) ctx,
                md_alg, hash, hash_len, sig, sig_len, f_rng, p_rng ) );
}
#endif
```
Have the user add there own `ecdsa_sign_wrap` methods.  This is the first entry point into the for signature so it make sense. Same goes for the verification process.

**Comment by RonEld** [comment]
@umanayana Thank you for your feedback.
However, I have to disagree with your analysis.
As mentioned before, `mbedtls_ecdsa_write_signature` is used to sign the data using the `mbedtls_ecdsa_sign()` function, and then encode the signature in an ASN1 notation.
The usage of `MBEDTLS_ECDSA_SIGN_ALT` is to replace `mbedtls_ecdsa_sign()` and that function  alone. Note that both `mbedtls_ecdsa_write_signature()` and `mbedtls_ecdsa_sign_det()` use the `mbedtls_ecdsa_sign()` intenally.
`mbedtls_ecdsa_sign()` does not receive `mbedtls_pk_context *` as parameter, and thus no need to replace it for your alternative implementation.
Once you replace `mbedtls_ecdsa_sign()`, then the hw acceleration for ecdsa is done "under the hood" for `mbedtls_ecdsa_write_signature()`.

From your previous questions, I understand that your use case is some secure element, so perhaps you need a better solution, such as an opaque key?

**Comment by umanayana** [comment]
@RonEld 

What is an opaque key? Can you provide some resource on this. 

I have moved back to the previous implementation where I have added my custom `mbedtls_ecdsa_sign()` and `mbedtls_ecdsa_verify()`. But I have run into a problem, see below

**Background information to understand question**
- My device has a total of 4 certificates.  Two server certificate, 1 device certificate and CA of the device certificate (ie the signer of the device certificate). 
- The hardware accelerator is only used for device certificate related stuff. That is signing and verify. This is because the hardware accelerator stores the public/private key pair in the hardware and they cannot be read. Well the public key can be read but not the private key.
- During the handshaking process the `mbedtls_ecdsa_verify()` is called during the verification process of the server certificates. But this will fail because the hardware accelerator cannot verify this certificate.

**Questions**
What is the best way to make sure the ecdsa signing and verification only occur when device certificate requires it?

I have some possible solution but all required modification of mbedTLS API.

Thanks

**Comment by umanayana** [comment]
Well one more update here got the verification process to work.

**Comment by RonEld** [comment]
@umanayana Thank you for your update.

An opaque key means some handle to the private key, that is known only in the secure element. Unfortunately, we don't have support for this yet. My previous question was to understand the need for this feature. I apologize for confusion.

>What is the best way to make sure the ecdsa signing and verification only occur when device certificate requires it?

I don't quite understand your question. In the TLS handshake, once a certificate is received from the remote peer, the certificate is being verified, according to the ciphersuite and public key of its issuer.

>Can I pass along some sort of context information some how so I can map a specific call to mbedtls_ecdsa_verify() to a given certificate?

>Can I pass along some sort of context information some how so I can map a specific call to mbedtls_ecdsa_verify() to a given certificate?

If I understand your question correct, it is up to the application to decide what key ( context)  to use when trying to verify a certificate. You can map in your application a context to a certificate, but it is already handled in the TLS handshake.

>Can I rebuild the 64byte public key from the `mbedtls_ecp_point` parameter?

The `mbedtls_ecp_point` is the public key. If you mean the binary format, I believe you mean the `mbedtls_ecp_point_write_binary` API. 

>I have some possible solution but all required modification of mbedTLS API.

It is unclear to me why you think Mbed TLS API should be modified. Unless I am misunderstanding your use case. 

In short, if your HW accelaration received the private \ public key in binary format instead of the `mbedtls_mpi`  and `mbedtls_ecp_point`, then your *alternative implementation* should convert these input parameters into the binary format, using `mbedtls_ecp_point_write_binary` and `mbedtls_mpi_write_binary` . You should consider your accelarator's byte order as well.

I am happy you managed to have your signing and verification process to work. I will close this question if no further action required, please update.

**Comment by umanayana** [comment]
@RonEld 
Thanks for the through explanation. What I was doing before was indeed wrong. After working through many errors I managed to understand the process better. 

Although now I have a ECDH problem, but this will be a new thread. So you can close this.

Thanks

**Comment by hanno-arm** [comment]
Thanks for the update, @umanayana. Closing.
